### PR TITLE
Allow to hook-in application-level logic to handle persistent references

### PR DESCRIPTION
**Pull request by navytux**
Continuing 84598fe1 (Add support for persistent references) theme let's
develop support for persistent references more: the usual situation (at
least in ZODB context, which is actually the original reason
persistent references exist) is that decode converts a reference of
(type, oid) tuple form into corresponding "ghost" object - an instance
of type, but not yet loaded with database object's data. This way
resulted object tree is created right away with types application
expects (and ZODB/py further cares to automatically load object data
when that ghost objects are accessed).

To mimic this on Go side, with current state, after decoding, one would
need to make a full reflect pass on the resulted decoded objects, find
Refs, and convert them to instances of other types, also caring to
patch pointers in other objects consistently that were pointing to such
Refs. In other words it is some work and it coincides almost 100% with
what Decoder already does by itself.

Thus, not to duplicate that work and conducting parallel with Python
world, let's add ability to configure Decoder and Encoder so that
persistent references could be handled with user-specified application
logic right in the process, and e.g. Decoder would return resulted
object with references converted to corresponding Go types.

To do so we introduce DecoderConfig and EncoderConfig - configurations
to tune decode/encode process. To maintain backward compatibility
NewDecoder and NewEncoder signatures are not adjusted and instead
NewDecoderWithConfig and NewEncoderWithConfig are introduces. We should
be sure we won't need e.g. NewDecoderWithConfigAndXXX in the future,
since from now on we could be adding fields to configuration structs
without breaking backward compatibility.

For decoding there is DecoderConfig.PersistentLoad which mimics
Unpickler.persistent_load in Python:

https://docs.python.org/3/library/pickle.html#pickle.Unpickler.persistent_load

For encoding there is EncoderConfig.PersistentID which mimics
Pickler.persistent_id in Python:

https://docs.python.org/3/library/pickle.html#pickle.Pickler.persistent_id

Then both Decoder and Encoder are adjusted correspondingly with tests
added.

About tests: I was contemplating to patch-in support for decoder and
encoder configurations, and handling errors, into our main test
TestDecode, but for now decided not to go that way and to put the test
as separate TestPersistentRefs.

By the way, with having configurations in place, we could start to add
other things there - e.g. the protocol version to use for Encoder.
Currently we have several places where we either don't use opcodes from
higher protocol versions (fearing to break compatibility), or instead
always use higher-version opcodes without checking we should be able to
do so by allowed protocol version.

### CODE REVIEW

File: encode.go [hunk 1]
```diff
@@ -20,12 +20,31 @@ func (te *TypeError) Error() string {
...
+type EncoderConfig struct {
+	// PersistentID, if !nil, will be used by encoder to encode objects as persistent references.
+	//
+	// Whenever encoders sees pointer to a Go struct object, it will call
+	// PersistentID to find out how to encode that object. If PersistentID
+	// returns nil, the object is encoded regularly. If !nil - the object
+	// will be encoded as Ref with persistent ID set to what was returned.
+	//
+	// See Ref documentation for details and further pointers on the subject.
+	PersistentID func(obj interface{}) interface{}
```
Comment (kisielk) [inline encode.go]: Would it make more sense to have this return `*Ref` instead? A `nil` value would still indicate to encode the object as normal.

File: ogorek.go [hunk 2]
```diff
@@ -123,10 +124,37 @@ type Decoder struct {
...
+	// PersistentLoad with reference's persistent ID. If PersistentLoad
+	// returns !nil object without error, decoder will use that object
+	// instead of Ref in the resulted built Go object.
+	//
+	// An example use-case for PersistentLoad is to transform persistent
+	// references in a ZODB database of form (type, oid) tuple, into
+	// equivalent-to-type Go ghost object, e.g. equivalent to zodb.BTree.
+	//
+	// See Ref documentation for details and further pointers on the subject.
+	PersistentLoad func(pid interface{}) (interface{}, error)
```
Comment (kisielk) [inline ogorek.go]: And then have this accept a `Ref` instead of an `interface{}` ?

Comment (navytux) [reply ogorek.go]: Thanks for feedback. I've actually started with PersistentLoad accepting Ref but than switched for it to accept Ref.Pid diectly, probably to mimic python api 1-1. On a second glance I agree that having less `interface{}` in the API would make things a bit more clear. Please let me know you preference - it should be easy for me to rework the patch to be

```
PersistentID func(obj interface{}) *Ref
PersistentLoad func(ref *Ref) (interface{}, error)
```

**Comment by navytux** [comment]
I've reworked it to be:

```
PersistentRef func(obj interface{}) *Ref
PersistentLoad func(ref Ref) (interface{}, error)
```

**Comment by navytux** [comment]
full interdiff:

```diff
diff --git a/encode.go b/encode.go
index 057f408..d648947 100644
--- a/encode.go
+++ b/encode.go
@@ -26,15 +26,15 @@ type Encoder struct {
 
 // EncoderConfig allows to tune Encoder.
 type EncoderConfig struct {
-	// PersistentID, if !nil, will be used by encoder to encode objects as persistent references.
+	// PersistentRef, if !nil, will be used by encoder to encode objects as persistent references.
 	//
 	// Whenever encoders sees pointer to a Go struct object, it will call
-	// PersistentID to find out how to encode that object. If PersistentID
+	// PersistentRef to find out how to encode that object. If PersistentRef
 	// returns nil, the object is encoded regularly. If !nil - the object
-	// will be encoded as Ref with persistent ID set to what was returned.
+	// will be encoded as returned Ref.
 	//
 	// See Ref documentation for details and further pointers on the subject.
-	PersistentID func(obj interface{}) interface{}
+	PersistentRef func(obj interface{}) *Ref
 }
 
 // NewEncoder returns a new Encoder struct with default values
@@ -97,10 +97,10 @@ func (e *Encoder) encode(rv reflect.Value) error {
 
 		if rv.Elem().Kind() == reflect.Struct {
 			// check if we have to encode this object as persistent reference.
-			if getpid := e.config.PersistentID; getpid != nil {
-				pid := getpid(rv.Interface())
-				if pid != nil {
-					return e.encodeRef(&Ref{Pid: pid})
+			if getref := e.config.PersistentRef; getref != nil {
+				ref := getref(rv.Interface())
+				if ref != nil {
+					return e.encodeRef(ref)
 				}
 			}
 
diff --git a/ogorek.go b/ogorek.go
index ad399b6..18636a9 100644
--- a/ogorek.go
+++ b/ogorek.go
@@ -129,16 +129,16 @@ type DecoderConfig struct {
 	// PersistentLoad, if !nil, will be used by decoder to handle persistent references.
 	//
 	// Whenever decoder finds one in pickle stream, it will call
-	// PersistentLoad with reference's persistent ID. If PersistentLoad
-	// returns !nil object without error, decoder will use that object
-	// instead of Ref in the resulted built Go object.
+	// PersistentLoad with the reference. If PersistentLoad returns !nil
+	// object without error, decoder will use that object instead of Ref in
+	// the resulted built Go object.
 	//
 	// An example use-case for PersistentLoad is to transform persistent
 	// references in a ZODB database of form (type, oid) tuple, into
 	// equivalent-to-type Go ghost object, e.g. equivalent to zodb.BTree.
 	//
 	// See Ref documentation for details and further pointers on the subject.
-	PersistentLoad func(pid interface{}) (interface{}, error)
+	PersistentLoad func(ref Ref) (interface{}, error)
 }
 
 // NewDecoder constructs a new Decoder which will decode the pickle stream in r.
@@ -505,7 +505,7 @@ func (d *Decoder) loadNone() error {
 //
 // See https://docs.python.org/3/library/pickle.html#pickle-persistent for details.
 //
-// See DecoderConfig.PersistentLoad and EncoderConfig.PersistentID for ways to
+// See DecoderConfig.PersistentLoad and EncoderConfig.PersistentRef for ways to
 // tune Decoder and Encoder to handle persistent references with user-specified
 // application logic.
 type Ref struct {
@@ -538,7 +538,7 @@ func (d *Decoder) loadBinPersid() error {
 // handleRef is common place to handle Refs.
 func (d *Decoder) handleRef(ref Ref) error {
 	if load := d.config.PersistentLoad; load != nil {
-		obj, err := load(ref.Pid)
+		obj, err := load(ref)
 		if err != nil {
 			return fmt.Errorf("pickle: handleRef: %s", err)
 		}
diff --git a/ogorek_test.go b/ogorek_test.go
index cd1fe96..2c0f33b 100644
--- a/ogorek_test.go
+++ b/ogorek_test.go
@@ -316,10 +316,10 @@ func TestPersistentRefs(t *testing.T) {
 
 	errInvalidRef := errors.New("invalid reference")
 
-	// Ref ID -> ? object
-	loadpid := func(pid interface{}) (interface{}, error) {
+	// Ref -> ? object
+	loadref := func(ref Ref) (interface{}, error) {
 		// pretend we handle "zodb.BTree" -> ZBTree.
-		t, ok := pid.(Tuple)
+		t, ok := ref.Pid.(Tuple)
 		if !ok || len(t) != 2 {
 			return nil, errInvalidRef
 		}
@@ -340,20 +340,20 @@ func TestPersistentRefs(t *testing.T) {
 		}
 	}
 
-	// object -> ? Ref ID
-	getpid := func(obj interface{}) interface{} {
+	// object -> ? Ref
+	getref := func(obj interface{}) *Ref {
 		// pretend we handle ZBTree.
 		switch obj := obj.(type) {
 		default:
 			return nil
 
 		case *ZBTree:
-			return Tuple{Class{Module: "zodb", Name: "BTree"}, obj.oid}
+			return &Ref{Pid: Tuple{Class{Module: "zodb", Name: "BTree"}, obj.oid}}
 		}
 	}
 
-	dconf := &DecoderConfig{PersistentLoad: loadpid}
-	econf := &EncoderConfig{PersistentID: getpid}
+	dconf := &DecoderConfig{PersistentLoad: loadref}
+	econf := &EncoderConfig{PersistentRef: getref}
 
 	testv := []struct {
 		input    string
```

**Comment by navytux** [comment]
hm, probably PersistentID -> PersistentRef renaming was not good  (maybe it should return *Ref, but still be named as PersistentID).

I'll pause for now to wait for your feedback. For the reference, we can go back to any version of the patch easily, or mix if needed.

**Comment by navytux** [comment]
> I think PersistentRef is better than PersistentID

ok

File: ogorek.go [hunk 3]
```diff
@@ -470,12 +498,16 @@ func (d *Decoder) loadNone() error {
 }
 
 
-// Ref represents Python's persistent reference.
+// Ref is default representation for a Python persistent reference.
```
[the remark this answers is not in the document]
Comment (navytux) [reply ogorek.go]: ok

File: ogorek.go [hunk 4]
```diff
@@ -123,10 +124,37 @@ type Decoder struct {
...
+	// Whenever decoder finds one in pickle stream, it will call
+	// PersistentLoad with the reference. If PersistentLoad returns !nil
+	// object without error, decoder will use that object instead of Ref in
+	// the resulted built Go object.
+	//
+	// An example use-case for PersistentLoad is to transform persistent
+	// references in a ZODB database of form (type, oid) tuple, into
+	// equivalent-to-type Go ghost object, e.g. equivalent to zodb.BTree.
+	//
+	// See Ref documentation for details and further pointers on the subject.
```
[the remark this answers is not in the document]
Comment (navytux) [reply ogorek.go]: ok (and in EncoderConfig too).

File: ogorek.go [hunk 5]
```diff
@@ -123,10 +124,37 @@ type Decoder struct {
...
+	PersistentLoad func(ref Ref) (interface{}, error)
+}
+
 // NewDecoder constructs a new Decoder which will decode the pickle stream in r.
 func NewDecoder(r io.Reader) *Decoder {
+	return NewDecoderWithConfig(r, &DecoderConfig{})
+}
+
+// NewDecoderWithConfig is similar to NewDecoder, but allows to specify decoder configuration.
```
[the remark this answers is not in the document]
Comment (navytux) [reply ogorek.go]: ok

File: ogorek.go [hunk 6]
```diff
@@ -123,10 +124,37 @@ type Decoder struct {
 	line  []byte
 }
 
+// DecoderConfig allows to tune Decoder.
+type DecoderConfig struct {
+	// PersistentLoad, if !nil, will be used by decoder to handle persistent references.
+	//
+	// Whenever decoder finds one in pickle stream, it will call
+	// PersistentLoad with the reference. If PersistentLoad returns !nil
+	// object without error, decoder will use that object instead of Ref in
```
[the remark this answers is not in the document]
Comment (navytux) [reply ogorek.go]: ok

File: ogorek.go [hunk 7]
```diff
@@ -123,10 +124,37 @@ type Decoder struct {
 	line  []byte
 }
 
+// DecoderConfig allows to tune Decoder.
+type DecoderConfig struct {
+	// PersistentLoad, if !nil, will be used by decoder to handle persistent references.
+	//
+	// Whenever decoder finds one in pickle stream, it will call
```
[the remark this answers is not in the document]
Comment (navytux) [reply ogorek.go]: ok

File: encode.go [hunk 8]
```diff
@@ -20,12 +20,31 @@ func (te *TypeError) Error() string {
...
+	PersistentRef func(obj interface{}) *Ref
 }
 
 // NewEncoder returns a new Encoder struct with default values
 func NewEncoder(w io.Writer) *Encoder {
-	return &Encoder{w: w}
+	return NewEncoderWithConfig(w, &EncoderConfig{})
+}
+
+// NewEncoderWithConfig is similar to NewEncoder, but allows to specify encoder configuration.
```
[the remark this answers is not in the document]
Comment (navytux) [reply encode.go]: ok

File: encode.go [hunk 9]
```diff
@@ -20,12 +20,31 @@ func (te *TypeError) Error() string {
...
+}
+
+// EncoderConfig allows to tune Encoder.
+type EncoderConfig struct {
+	// PersistentRef, if !nil, will be used by encoder to encode objects as persistent references.
+	//
+	// Whenever encoders sees pointer to a Go struct object, it will call
+	// PersistentRef to find out how to encode that object. If PersistentRef
+	// returns nil, the object is encoded regularly. If !nil - the object
+	// will be encoded as returned Ref.
```
[the remark this answers is not in the document]
Comment (navytux) [reply encode.go]: ok

File: encode.go [hunk 10]
```diff
@@ -20,12 +20,31 @@ func (te *TypeError) Error() string {
...
-	w io.Writer
+	w      io.Writer
+	config *EncoderConfig
+}
+
+// EncoderConfig allows to tune Encoder.
+type EncoderConfig struct {
+	// PersistentRef, if !nil, will be used by encoder to encode objects as persistent references.
+	//
+	// Whenever encoders sees pointer to a Go struct object, it will call
```
[the remark this answers is not in the document]
Comment (navytux) [reply encode.go]: ok

**Comment by navytux** [comment]
Patch amended with the following interdiff:

```diff
diff --git a/encode.go b/encode.go
index d648947..0f4bfd3 100644
--- a/encode.go
+++ b/encode.go
@@ -28,12 +28,12 @@ type Encoder struct {
 type EncoderConfig struct {
 	// PersistentRef, if !nil, will be used by encoder to encode objects as persistent references.
 	//
-	// Whenever encoders sees pointer to a Go struct object, it will call
+	// Whenever the encoders sees pointer to a Go struct object, it will call
 	// PersistentRef to find out how to encode that object. If PersistentRef
 	// returns nil, the object is encoded regularly. If !nil - the object
-	// will be encoded as returned Ref.
+	// will be encoded as an object reference.
 	//
-	// See Ref documentation for details and further pointers on the subject.
+	// See Ref documentation for more details.
 	PersistentRef func(obj interface{}) *Ref
 }
 
@@ -42,7 +42,7 @@ func NewEncoder(w io.Writer) *Encoder {
 	return NewEncoderWithConfig(w, &EncoderConfig{})
 }
 
-// NewEncoderWithConfig is similar to NewEncoder, but allows to specify encoder configuration.
+// NewEncoderWithConfig is similar to NewEncoder, but allows specifying the encoder configuration.
 func NewEncoderWithConfig(w io.Writer, config *EncoderConfig) *Encoder {
 	return &Encoder{w: w, config: config}
 }
diff --git a/ogorek.go b/ogorek.go
index 18636a9..8e55bb1 100644
--- a/ogorek.go
+++ b/ogorek.go
@@ -128,16 +128,16 @@ type Decoder struct {
 type DecoderConfig struct {
 	// PersistentLoad, if !nil, will be used by decoder to handle persistent references.
 	//
-	// Whenever decoder finds one in pickle stream, it will call
-	// PersistentLoad with the reference. If PersistentLoad returns !nil
-	// object without error, decoder will use that object instead of Ref in
+	// Whenever the decoder finds an object reference in the pickle stream
+	// it will call PersistentLoad. If PersistentLoad returns !nil object
+	// without error, the decoder will use that object instead of Ref in
 	// the resulted built Go object.
 	//
 	// An example use-case for PersistentLoad is to transform persistent
 	// references in a ZODB database of form (type, oid) tuple, into
 	// equivalent-to-type Go ghost object, e.g. equivalent to zodb.BTree.
 	//
-	// See Ref documentation for details and further pointers on the subject.
+	// See Ref documentation for more details.
 	PersistentLoad func(ref Ref) (interface{}, error)
 }
 
@@ -146,7 +146,7 @@ func NewDecoder(r io.Reader) *Decoder {
 	return NewDecoderWithConfig(r, &DecoderConfig{})
 }
 
-// NewDecoderWithConfig is similar to NewDecoder, but allows to specify decoder configuration.
+// NewDecoderWithConfig is similar to NewDecoder, but allows specifying decoder configuration.
 func NewDecoderWithConfig(r io.Reader, config *DecoderConfig) *Decoder {
 	reader := bufio.NewReader(r)
 	return &Decoder{
@@ -498,7 +498,7 @@ func (d *Decoder) loadNone() error {
 }
 
 
-// Ref is default representation for a Python persistent reference.
+// Ref is the default representation for a Python persistent reference.
 //
 // Such references are used when one pickle somehow references another pickle
 // in e.g. a database.
```

**Comment by navytux** [comment]
(I'm not sure how to mark you reviews as resolved to remove the red cross; please forgive me lack of github fu a bit on my side)

**Comment by navytux** [comment]
(for the reference e.g. Gerrit allows to see interdiff in between patch versions easily in its UI, and there is clearly visible "done" button to mark a review comment as addressed)

**Comment by kisielk** [comment]
In github you just add more commits to your branch, instead of squashing, then you can view the diff between each commit. I think you can then also click done on the requested changes. It's quite different from gerrit workflow.

**Comment by navytux** [comment]
Searching finds things like https://github.com/isaacs/github/issues/468 and https://github.com/isaacs/github/issues/251 so I guess this is not implemented.

I understand about github workflow. There are two histories: short-term history which is useful while a patch is being developed - to see how it changes from its revision from revision. And long-term history which is the git history of a project. If we let intermediate draft edits to propagate into long-term history the long-term  history starts to loose its value - the ability to learn about codebase from e.g. git blame. For this reason I try to keep one change that does one thing (and hopefully well) as only one patch, and then workaround lack of short-term history with posting interdiff in comments.

Thanks for merging the patch.

Kirill

**Comment by kisielk** [comment]
That's where the "squash and merge" feature of Github comes in handy. It takes all that short-term history and squashes it down to one commit before creating the merge.

**Comment by navytux** [comment]
Yes, sure, but it depends on the maintainer to actually apply it. And there is no guarantee that it will be applied, nor that with this approach the patch description itself could be iteratively amended to desired text (with "squash and merge" it will just concatenate text from all draft commits).

**Comment by kisielk** [comment]
Of course, but ultimately it's the maintainer's repo so it's up to them how they want to manage it or if they care about those kinds of things :)

**Comment by navytux** [comment]
Yes, sure. I was just referencing it all for completeness. If you insist on having draft commits added on top of the main one, I will do it this way next time, but if it is not a hard requirement I will keep on going with my amend way.

Thanks again for applying the patch.

**Comment by kisielk** [comment]
I've fine with whatever you prefer, I was just mentioning that what you described was possible.

**Comment by navytux** [comment]
I see, thanks.

**Pull request merged**